pep.rst

```diff
@@ -519,6 +533,62 @@ Likewise:
 
     mytag()
 
+HTML Example of Rich Return Types
+=================================
+
+Tag functions can be a powerful part of larger processing chains by returning richer objects.
+JavaScript tagged template literals, for example, are not constrained by a requirement to
+return a string. As an example, let's look at an HTML generation system, with a usage and
+"subcomponent":
+
+.. code-block::
+
+    def Menu(icon: str) -> HTML:
+        return html'<img alt="Site Logo" src={icon} />'
+
+    icon = 'acme.png'
+    result = html'<header><{Menu} logo={icon} class="my-menu"/></header>'
+    img = result.children[0]
+    assert img.tag == "img"
+    assert img.props == {"logo": "acme.png", "class": "my-menu"}
```

Comment:
I'm having a hard time understanding this example without an example implementation of the `html` tag. I've used JSX extensively, but this is not intuitive for me and I can't pinpoint exactly why. I guess my question is how did the `tag` change to `img`, but the props stayed `logo` and `class`. And did `alt` and `src` get lost?

Also, it should probably be `img.tagname` and `img.attrs` according to the protocal below.

Reply:
> I'm having a hard time understanding this example without an example implementation of the html tag

Yes, this was what I tried to highlight, too &mdash; sorry if I didn't do a good job. 

For instance, it is current form, what is `str(result)` expected to be here? Specifically, what is the value for the `src=` attribute of the `img` tag going to be? (The `html` tag can do anything, of course, but this example sort of breaks expectations/intuition from JSX-land.)

On the other hand, this would make sense to me and (I think?) to other devs who have used JSX in the past:

```python
def Menu(*, logo: str, class: str) -> HTML:
    return html'<img alt="Site Logo" src={logo} class={class} />'
    
icon = 'acme.png'
result = html'<header><{Menu} logo={icon} class="my-menu"/></header>'
img = result.children[0]
assert img.tagname == "img"
assert img.attrs == {"src": "acme.png", "class": "my-menu"}
```

Reply:
You are both correct. I went too fast on committing this, adding in `class` on one side and not the other and not getting `src` into the `attrs`. My apologies. I will paste what Dave has. Extra credit for keyword-only arguments.

Are you ok with leaving in the `str()` part? The hope was, a usage that showed 2 ways to use the return type: having the data in a VDOM like thing, but also converting to a string for HTML serialization.

Reply:
> Are you ok with leaving in the str() part? The hope was, a usage that showed 2 ways to use the return type: having the data in a VDOM like thing, but also converting to a string for HTML serialization.

Yes! Personally, I like this. :-)

(I just re-scaned the PEP and don't see an explicit example with `str(...)` and `HTML`?)

Reply:
I meant from conversation. Something like:

```python
# We can also treat the return type as a string of specially-serialized HTML
assert str(result) = '<header>' # etc.
```

...as the last assert.

Reply:
Gotcha. Yes, I personally like that; it matches what I think I'd want out of an _actual_ `html` tag.

Reply:
This subsection should be removed, now that we allow dotted tag names.
https://github.com/jimbaker/tagstr/blob/pe-process-nits/pep.rst#no-dotted-tag-names

Reply:
Not sure I see your point. Which subsection relates to dotted tag names?

Reply:
Jim was right, another short section still said no dotted. I just removed it.

I also processed (I believe) everything in this conversation and pushed. Marking as resolved.